[WIP-topotest] bgpd: handle link-local for vpnv4 NLRI with V6 nexthop

samanvithab opened the pull request:
As per RFC5549 Section 3,
This document extends the definition of the AFI/SAFI for advertisement of IPv4 NLRI
   and VPN-IPv4 NLRI to extend the set of network-layer protocols to
   which the Next Hop address can belong, to include IPv6 in addition to
   IPv4.
   Specifically, this document allows advertising with [RFC4760] of an
   MP_REACH_NLRI with:
   -  AFI = 1
   -  SAFI = 1, 2, 4, or 128
   -  Length of Next Hop Address = 16 or 32
   -  Next Hop Address = IPv6 address of next hop (potentially followed by the link-local IPv6 address of the next hop).  This field is to be constructed as per Section 3 of [RFC2545].

Issue:
On a point to point link, where the peers share common subnet, vpnv4 NLRIs advertised does not carry link-local address in nexthop field when extended nexthop capability is configured on the peer.

Fix:
Nexthop field to be constructed as per Section 3 of [RFC2545] for AFI-SAFI <1,128> when extended nexthop capability is configured on the peer.

ton31337 commented on bgpd/bgp_route.c:
```diff
@@ -2331,7 +2331,8 @@ bool subgroup_announce_check(struct bgp_dest *dest, struct bgp_path_info *pi,
 	 || ((safi == SAFI_ENCAP || safi == SAFI_MPLS_VPN)                     \
-	     && attr->mp_nexthop_len >= IPV6_MAX_BYTELEN))
+	     && attr->mp_nexthop_len >= IPV6_MAX_BYTELEN)                      \
+	 || (safi == SAFI_MPLS_VPN && peer_cap_enhe(peer, afi, safi)))
```
Hmm, I don't understand how is this related to `NEXTHOP_IS_V6`. This condition `(safi == SAFI_MPLS_VPN && peer_cap_enhe(peer, afi, safi))` doesn't say that this is IPv6 next-hop or I don't understand something? To me it sounds like we should check `attr->mp_nexthop_len >= BGP_ATTR_NHLEN_VPNV4`...

samanvithab replied:
RFC5549 supports IPV4/VPNV4 NLRIs to be advertised with IPV6 nexthop to the BGP peer when extended nexthop encoding capability is enabled.
peer_cap_enhe(peer, afi, safi) returns true if ‘extended nexthop encoding capability’ is enabled for the <afi,safi> as per RFC5549. This capability is configured via below command,
neighbor <> capability extended-nexthop

Currently, the mentioned issue case is working fine for IPV4 NLRI advertised with IPV6 nexthop & issue is only with VPNv4 NLRI.

If we consider the working IPV4 case when extended nexthop encoding capability enabled on peer, 
for a IPV4 route in the RIB with V4 nexthop, the attr->mp_nexthop_len is 4. 
But, the first check in NEXTHOP_IS_V6 will return true as peer_cap_enhe(peer, afi, safi) is true. 
 >>>  #define NEXTHOP_IS_V6
		(safi != SAFI_ENCAP && safi != SAFI_MPLS_VPN  && (p->family == AF_INET6 || peer_cap_enhe(peer, afi, safi))

While announcing the route, based on this NEXTHOP_IS_V6 check we set the appropriate IPV6 nexthop length as 16 or 32 with LL if they share common subnet. 

But the same handling is missing for VPNV4 case.  NEXTHOP_IS_V6 check returns false as mp_nexthop_len is 4 & safi SAFI_MPLS_VPN.
Currently for VPNv4 case with extended nexthop capability enabled, IPV6 nexthop length is being updated at bgp_packet_mpattr_start(), where nh_afi is set based on the peer_cap_enhe() check. Further, based on this nh_afi only global IP is being added as nexthop, LL is not getting added.

In bgp_packet_mpattr_start(),

        /* Nexthop AFI */
        if (afi == AFI_IP
            && (safi == SAFI_UNICAST || safi == SAFI_LABELED_UNICAST
                || safi == SAFI_MPLS_VPN || safi == SAFI_MULTICAST))
                nh_afi = peer_cap_enhe(peer, afi, safi) ? AFI_IP6 : AFI_IP;


So, this seems to be the right thing to do considering similar handling for IPV4 case in Line No 2332.

ton31337 replied:
>RFC5549 supports IPV4/VPNV4 NLRIs to be advertised with IPV6 nexthop to the BGP peer when extended nexthop encoding capability is enabled.

I don't say it's wrong :smile: I just say, that NEXTOP_IS_V6 macro is not the right place to modify here like it's now. As I suggested, wouldn't it be enough checking `mp_nexthop_len` properly?

samanvithab commented:
CI:rerun “Failure not related to my changes”

samanvithab commented:
CI:rerun “Failure not related to my changes”

ton31337 commented:
Can we add a topotest for verify this?

samanvithab committed:
bgpd: handle link-local for vpnv4 NLRI with V6 nexthop

As per RFC5549 Section 3,
This document extends the definition of the AFI/SAFI for advertisement of IPv4 NLRI
and VPN-IPv4 NLRI to extend the set of network-layer protocols to
which the Next Hop address can belong, to include IPv6 in addition to
IPv4.
Specifically, this document allows advertising with [RFC4760] of an
MP_REACH_NLRI with:

AFI = 1
SAFI = 1, 2, 4, or 128
Length of Next Hop Address = 16 or 32
Next Hop Address = IPv6 address of next hop (potentially followed by the link-local IPv6 address of the next hop). This field is to be constructed as per Section 3 of [RFC2545].
Issue:
On a point to point link, where the peers share common subnet, vpnv4 NLRIs advertised does not carry link-local address in nexthop field when extended nexthop capability is configured on the peer.

Fix:
Nexthop field to be constructed as per Section 3 of [RFC2545] for AFI-SAFI <1,128> when extended nexthop capability is configured on the peer.

Signed-off-by: Samanvitha B Bhargav <samanvitha-b.bhargav@broadcom.com>

samanvithab committed:
tests: Added topotests for RFC5549 to test VPNv4 with V6 nexthop

Author: Naveen Guggarigoudar <naveen.guggarigoudar@broadcom.com>
Signed-off-by: Samanvitha B Bhargav <samanvitha-b.bhargav@broadcom.com>

riw777 commented:
@frrbot autoclose in 2 weeks
